lightning/src/util/persist.rs

```diff
@@ -64,41 +114,387 @@ impl<'a, A: KVStorePersister, M: Deref, T: Deref, ES: Deref, NS: Deref, SP: Dere
+	) -> io::Result<(BlockHash, ChannelMonitor<<SP::Target as SignerProvider>::Signer>)>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized;
+	/// List all the names of updates corresponding to a given monitor name.
+	fn list_update_names(&self, monitor_name: &MonitorName) -> io::Result<Vec<UpdateName>>;
+	/// Path to corresponding update directory for a given monitor name.
+	fn path_to_monitor_updates(&self, monitor_name: &MonitorName) -> String;
+	/// Deserialize a channel monitor update.
+	fn deserialize_monitor_update(
+		&self, monitor_name: &MonitorName, update_name: &UpdateName,
+	) -> io::Result<ChannelMonitorUpdate>;
+	/// Key to a specific update.
+	fn update_key(&self, monitor_name: &MonitorName, update_name: &UpdateName) -> String;
+	/// Delete updates with an update_id lower than the given channel monitor.
+	fn delete_stale_updates<ChannelSigner: WriteableEcdsaChannelSigner>(
+		&self, channel_id: OutPoint, monitor: &ChannelMonitor<ChannelSigner>,
+	) -> io::Result<()>;
+}
+
+impl<K: KVStore> KVStoreChannelMonitorReader<K> for K {
+	fn read_channelmonitors<
+		ES: Deref + Clone,
+		SP: Deref + Clone,
+		B: Deref,
+		F: Deref + Clone,
+		L: Deref,
+	>(
+		&self, entropy_source: ES, signer_provider: SP, broadcaster: &B, fee_estimator: F,
+		logger: &L,
+	) -> std::io::Result<Vec<(BlockHash, ChannelMonitor<<SP::Target as SignerProvider>::Signer>)>>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+		B::Target: BroadcasterInterface,
+		F::Target: FeeEstimator,
+		L::Target: Logger
+	{
+		let mut res = Vec::new();
+		// for each monitor...
+		for monitor_name in self.list_monitor_names()? {
+			// ...parse the monitor
+			let (bh, monitor) = self.deserialize_monitor(
+				entropy_source.clone(),
+				signer_provider.clone(),
+				monitor_name.clone(),
+			)?;
+			// ...parse and apply the updates with an id higher than the monitor.
+			for update_name in self.list_update_names(&monitor_name)? {
+				let update = self.deserialize_monitor_update(&monitor_name, &update_name)?;
+				if update.update_id == CLOSED_CHANNEL_UPDATE_ID
+					|| update.update_id > monitor.get_latest_update_id()
+				{
+					monitor
+						.update_monitor(&update, broadcaster, fee_estimator.clone(), logger)
+						.map_err(|_| {
+							KVStoreChannelMonitorReaderError::UpdateFailed(
+								"update_monitor returned Err(())".to_string(),
+								format!("monitor: {:?}", monitor_name),
+							)
+						})?;
+				}
+			}
+			// ...push the result into the return vec
+			res.push((bh, monitor))
+		}
+		Ok(res)
+	}
+
+	/// Key to a specific monitor.
+	fn monitor_key(&self, monitor_name: &MonitorName) -> String {
+		CHANNEL_MONITOR_PERSISTENCE_NAMESPACE.to_owned() + &monitor_name.0
+	}
+
+	/// Key to a specific update.
+	fn update_key(&self, monitor_name: &MonitorName, update_name: &UpdateName) -> String {
+		self.path_to_monitor_updates(monitor_name) + &update_name.0
```

Comment:
Ah. Yeah this comment reminds me. I took out `PathBuf` and just did concatenation, but I'll need to intersperse `/`.

But that aside -- yes, how do we want to handle depth in namespaces? I could do this other ways that are flatter, probably. I think since many users will at least start with a filesystem, a specific path for the updates for a monitor is important as it makes listing the monitor's updates easier/faster. That's not just a startup thing, we have to do that for every new block to write the consolidated monitor.

Reply:
I guess, at the the end of the day, if I want more than two levels (root + 1 namespace), I can just do it by handling delimiters in my backend the same way I would/do now. So, it's fine.

Using this in a way that I _think_ is idiomatic now does mean that the breadth of namespaces ~= number of channels.